### DISCUSSION TITLE
Can we add a protected IChildDataPortal<C> GetChildDataPortal<C>()

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#2838] [Ideas]
Currently, it looks like we need to cast to `IChildDataPortal<C>` so we can call `CreateChild()`:
```cs
var childDataPortal = (IChildDataPortal<Child>)GetDataPortal<Child>();

ChildInstance       = childDataPortal.CreateChild();
```

Without this when calling `FieldManager.UpdateChildren(this)` an exception is thrown by:
```cs
public ServiceProviderMethodInfo FindDataPortalMethod<T>(Type targetType, object[] criteria, bool throwOnError = true)
```
as it's looking for `[Insert]/[Update]` instead of `[InsertChild]/[UpdateChild]`.

**Comment by rockfordlhotka** [comment]
In what context are you interacting with the child data portal (outside of the logical server-side data portal operation methods)?

**Reply by michaelcsikos** [reply]
I sometimes have a child object which is optional, so in the parent object there's a `bool HasChild` property, which when toggled either creates the child or sets it `null`:

```cs
 public static readonly PropertyInfo<bool> HasChildProperty
                  = RegisterProperty<bool>(nameof(HasChild), StringsBiz.Child);
 public bool HasChild
 {
     get { return ReadProperty(HasChildProperty); }
     set { SetProperty(HasChildProperty, value); }
 }

 public static readonly PropertyInfo<Child> ChildProperty
                  = RegisterProperty<Child>(nameof(Child), StringsBiz.Child);
 public Child Child
 {
     get => GetProperty(ChildProperty);
     private set => SetProperty(ChildProperty, value);
 }

 private void OnHasChildChanged()
 {
     if (HasChild)
     {
         if (!FieldManager.FieldExists(ChildProperty) || Child == null)
         {
             var childDataPortal = (IChildDataPortal<Child>)GetDataPortal<Child>();

             Child               = childDataPortal.CreateChild(Id);
         }
     }
     else
         Child = null;
 }

 protected override void OnPropertyChanged(IPropertyInfo propertyInfo)
 {
     base.OnPropertyChanged(propertyInfo);

     if (propertyInfo == HasChildProperty)
         OnHasChildChanged();
 }
```

So, this is happening on the client, I suppose.

**Reply by rockfordlhotka** [reply]
Right, on the client, so technically calling the child data portal on the client isn't a valid scenario.

**Reply by michaelcsikos** [reply]
So how do we construct a new child on demand? In most cases it doesn't require a round-trip, it's just creating a new instance. The reason I don't want the child object to exist all the time is primarily because of the potential for broken rules even after the option is switched off. Setting the child back to null is the easiest way to ensure no broken rules are hanging around.

**Reply by rockfordlhotka** [reply]
I fully understand what you are saying. At the same time, in a data portal operation method you should expect to be able to interact with the data access layer - and that expectation fails if the code runs in arbitrary locations. You and I might understand the difference, but the wider user base of CSLA would surely struggle with that lack of consistency.

If a child needs to be created "out of band" (not during a parent's create/fetch operation) the model is to implement a creator/getter type. This is a root type so you can use the client-side data portal as normal. The `RunLocal` attribute can be used if no access to the physical server is required.

```c#
[Serializable]
public class ChildFactory : ReadOnlyBase<ChildFactory>
{
  // implement Child property as normal

  [Fetch]
  [RunLocal]
  private void Fetch([Inject] IChildDataPortal<Child> portal)
  {  
    Child = portal.Fetch();
  }
}
```

**Reply by michaelcsikos** [reply]
Thanks @rockfordlhotka. For the sake of anyone else who has the same question, you can also add `internal static` methods to ensure you are passing the correct parameters:
```cs
#region Factory

 internal static Child NewChild(IDataPortal<Factory> dataPortal, Guid id)
 {
     var factory = dataPortal.Create(id);

     return factory.Child;
 }

 internal static async Task<Child> GetChildAsync(IChildDataPortal<Child> childDataPortal, Guid id)
 {
     return await childDataPortal.FetchChildAsync(id);
 }

 [Serializable]
 internal class Factory
     : ReadOnlyBase<Factory>
 {
     public static readonly PropertyInfo<Child> ChildProperty = RegisterProperty<Child>(nameof(Child));
     public Child Child
     {
         get => ReadProperty(ChildProperty);
         private set => LoadProperty(ChildProperty, value);
     }

     [Create]
     [RunLocal]
     private void Create(Guid id, [Inject] IChildDataPortal<Child> childDataPortal)
     {
         Child = childDataPortal.CreateChild(id);
     }
 }

#endregion
```
Then from the parent:
```cs
var dataPortal = GetDataPortal<Child.Factory>();

ChildInstance  = Child.NewChild(dataPortal, Id);
```
